tests/unit/lms/services/course_test.py

```diff
@@ -44,15 +46,15 @@ def test_it_does_nothing_if_theres_already_a_matching_row(
 
     @pytest.mark.parametrize("canvas_sections_enabled", [True, False])
     def test_it_inserts_False_if_theres_a_matching_row_in_course_groups_exported_from_h(
-        self, ai_getter, db_session, svc, canvas_sections_enabled
+        self, application_instance_service, db_session, svc, canvas_sections_enabled
     ):
         db_session.add(
             CourseGroupsExportedFromH(
                 authority_provided_id="test_authority_provided_id",
                 created=datetime.datetime.utcnow(),
             )
         )
-        ai_getter.settings.return_value.set(
+        application_instance_service.get.return_value.settings.set(
             "canvas", "sections_enabled", canvas_sections_enabled
         )
 
```

Comment:
The `svc` fixture could be changed to instantiate the service directly rather than through a factory. But the service instance would still be a fixture. If each individual test instantiated the service itself, then every test would have to do this:

```python
def test_foo(self, pyramid_request)
    svc = ApplicationInstanceService(pyramid_request.db, "test_consumer_key")
    ...
```

For services that need more arguments (things from the request, mocks of other services, ...) this can get a lot worse, as each test could have to depend on several fixtures (mock services etc) and make a complex service class `__init__()` call.

And if the service's arguments changed every test would have to be updated.

Reply:
Yeah, I'm just suggesting making the `svc` fixture instantiate it directly (not removing the fixture), and have separate tests for the factory, precisely because this would make it less end-to-end and more unit-y. If the service arguments change, you just change the one fixture.

This prevents the integration test type situation where you make one mistake in factory and every test blows up. This way only the factory tests would blow up, which would give you the strong hint it's the factory, not the class that's at fault.

Checkmate is a pretty good example of this style: https://github.com/hypothesis/checkmate/blob/main/tests/unit/checkmate/services/secure_link_test.py